metagraph/src/common/algorithms.hpp

```diff
@@ -66,37 +66,93 @@ namespace utils {
         return mask;
     }
 
-    template <class AIt, class BIt>
-    uint64_t count_intersection(AIt first_begin, AIt first_end,
-                                BIt second_begin, BIt second_end) {
-        assert(std::is_sorted(first_begin, first_end));
-        assert(std::is_sorted(second_begin, second_end));
-        assert(std::set<typename AIt::value_type>(first_begin, first_end).size()
-                    == static_cast<uint64_t>(std::distance(first_begin, first_end)));
-        assert(std::set<typename BIt::value_type>(second_begin, second_end).size()
-                    == static_cast<uint64_t>(std::distance(second_begin, second_end)));
+    template <class InIt1, class InIt2>
+    constexpr uint64_t count_intersection(InIt1 a_begin, InIt1 a_end, InIt2 b_begin, InIt2 b_end) {
+        assert(std::is_sorted(a_begin, a_end));
+        assert(std::is_sorted(b_begin, b_end));
+        assert(std::adjacent_find(a_begin, a_end) == a_end);
+        assert(std::adjacent_find(b_begin, b_end) == b_end);
 
         uint64_t count = 0;
 
-        while (first_begin != first_end && second_begin != second_end) {
-            first_begin = std::lower_bound(first_begin, first_end, *second_begin);
+        while (a_begin != a_end && b_begin != b_end) {
+            if (*a_begin < *b_begin) {
+                a_begin = std::lower_bound(a_begin, a_end, *b_begin);
+            } else if (*b_begin < *a_begin) {
+                b_begin = std::lower_bound(b_begin, b_end, *a_begin);
```

Comment:
If this is called a lot, would be good to check if linear search is indeed not faster.

Reply:
I guess it's very input-dependent...I'll double-check gcc's implementation to see if we can gain any insights from it.

Reply:
hmm, looks like the implementation in libstdc++ does a linear search
https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/stl_algo.h#L5241

Reply:
same with clang

Reply:
yeah, I think that should generally work faster, unless you have very large sets with a small intersection.